assets/src/edit-story/components/form/advancedDropDown/container.js

```diff
@@ -0,0 +1,140 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import { useCallback, useEffect, useRef, useState } from 'react';
+import PropTypes from 'prop-types';
+import styled from 'styled-components';
+/**
+ * Internal dependencies
+ */
+import useFocusOut from '../../../utils/useFocusOut';
+import List from './list';
+import SearchInput from './searchInput';
+import { isKeywordFilterable } from './utils';
+
+const Container = styled.div`
+  position: relative;
+  display: flex;
+  flex-direction: row;
+  flex-wrap: wrap;
+  width: 100%;
+  min-width: 160px;
+  z-index: 2;
+  background-color: ${({ theme }) => theme.colors.bg.black};
+  border-radius: ${({ theme }) => theme.border.radius.default};
+  padding: 5px;
+  margin-top: 16px;
+`;
+
+function OptionsContainer({
+  value,
+  onSelect,
+  onClose,
+  isOpen,
+  getOptionsByQuery,
+  hasSearch,
+  onObserve,
+  options,
+  primaryOptions,
+  primaryLabel,
+  priorityOptions,
+  priorityLabel,
+  renderer,
+}) {
+  const ref = useRef();
+  const inputRef = useRef();
+  const [searchKeyword, setSearchKeyword] = useState('');
+  const [queriedOptions, setQueriedOptions] = useState(null);
+  const [isExpanded, setIsExpanded] = useState(true);
+  const [trigger, setTrigger] = useState(0);
+
+  useFocusOut(ref, onClose, [onClose]);
+
+  const handleSearchInputChanged = useCallback(({ target }) => {
+    setSearchKeyword(target.value);
+  }, []);
+
+  const handleLoadOptions = useCallback(() => {
+    getOptionsByQuery(searchKeyword).then((result) => {
+      setQueriedOptions(result);
+    });
+  }, [getOptionsByQuery, searchKeyword]);
+
+  useEffect(() => {
+    if (getOptionsByQuery && isKeywordFilterable(searchKeyword)) {
+      handleLoadOptions();
+    } else {
+      setQueriedOptions(null);
+    }
+  }, [getOptionsByQuery, searchKeyword, handleLoadOptions]);
+
+  useEffect(() => {
+    if (isOpen) {
+      inputRef?.current?.focus();
+    }
+  }, [isOpen]);
+
+  return (
+    <Container role="dialog" ref={ref}>
+      {hasSearch && (
+        <SearchInput
+          ref={inputRef}
+          value={searchKeyword}
+          onChange={handleSearchInputChanged}
+          onClose={onClose}
+          isExpanded={isExpanded}
+          focusFontListFirstOption={() => setTrigger((v) => ++v)}
+        />
+      )}
+      <List
+        value={value}
+        keyword={searchKeyword}
+        onSelect={onSelect}
+        onClose={onClose}
+        onExpandedChange={setIsExpanded}
+        focusTrigger={trigger}
+        onObserve={onObserve}
+        options={options || queriedOptions}
+        primaryOptions={primaryOptions}
+        primaryLabel={primaryLabel}
+        priorityOptions={priorityOptions}
+        priorityLabel={priorityLabel}
+        renderer={renderer}
+      />
```

Comment:
I really hate prop drilling like this (most of these props pass through the `OptionsContainer` component without being used). Could we perhaps use a render prop here, and just pass the props in the other direction?

So that would mean that this component can remove a lot of props and this render becomes:

```js
{children({searchKeyword,setIsExpanded,trigger,queriedOptions})}
```

And the parent invocation then changes from this:

```js
          <OptionsContainer
            isOpen={isOpen}
            value={selectedId}
            onSelect={handleSelect}
            onClose={debouncedCloseDropDown}
            getOptionsByQuery={getOptionsByQuery}
            hasSearch={hasSearch}
            onObserve={onObserve}
            options={options}
            primaryOptions={primaryOptions}
            primaryLabel={primaryLabel}
            priorityOptions={priorityOptions}
            priorityLabel={priorityLabel}
            renderer={renderer}
          />
```

To this:

```js
          <OptionsContainer
            isOpen={isOpen}
            onClose={debouncedCloseDropDown}
            getOptionsByQuery={getOptionsByQuery}
          >{({searchKeyword,setIsExpanded,trigger,queriedOptions}) => (
              <List
                value={value}
                keyword={searchKeyword}
                onSelect={onSelect}
                onClose={onClose}
                onExpandedChange={setIsExpanded}
                focusTrigger={trigger}
                onObserve={onObserve}
                options={options || queriedOptions}
                primaryOptions={primaryOptions}
                primaryLabel={primaryLabel}
                priorityOptions={priorityOptions}
                priorityLabel={priorityLabel}
                renderer={renderer}
              />)}</OptionsContainer>
```

I feel like that's a better result with a clearer separation of responsibility between the two classes. Makes sense?

Reply:
Used `renderContent` prop instead since we're doing this elsewhere as well instead of using `children` directly. Perhaps it's easier to read this way.